tests/diagnostics/non-void-return-constructor.slang

```diff
@@ -0,0 +1,47 @@
+//TEST:SIMPLE(filecheck=CHECK):
+
+// Test to catch compilation time errors with return values in a constructor / member functions within a class / struct.
+struct MyType
+{
+    float v;
+ 
+    // A constructor from the callsite's point of view is a function that returns MyType.
+    // A constructor from the inside the body is treated as a function that modifies `this`.
+    // because we are expected to write:
+    // this.v = 5;
+
+    // A test case where the constructor returns a value.
+    // This automatically gets resolved as this = t; return;
+    __init()
+    {
+        MyType t;
+        t.v = 5;
+        return t;
+    }
```

Comment:
Shouldn't this be an error?
Why do we allow this?

Reply:
there has been test cases where something like this was done, and as a result, we wanted to fix it within the IR lowering, and not complain.
@csyonghe , should we just fix the Falcor test, and have the error thrown, or proceed with the silent fix?

Reply:
Let's try to support this without throwing an error, because this is semantically valid.

`__init` is treated at the IR level something that returns a newly constructed value, instead of a function that modifies an existing value. However, this is not true for `[NonCopyableType]`, in which case we are only modifying `this`. See `slang-lower-to-ir.cpp, maybeAddReturnDestinationParam`.

Therefore, the right fix should be:
If we see `return ...`, translate it into `this = ...; return` if we are lowering to NonCopyableType, or translate it into `this = ...; return *this; ` if we are lowering for other types.

We can tell if we are lowering a constructor for a `NonCopyableType` by checking `context->returnDestination.flavor != LoweredValInfo::Flavor::None`. (We already do this in `visitReturnStmt`.